### PR TITLE
feedback

### PR DESCRIPTION
**Pull request by dreamfall**

### CODE REVIEW

File: app/concerns/controllers/feedback_mixin.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
+module Controllers::FeedbackMixin
+  extend ActiveSupport::Concern
+
+  def self.included(base)
+    base.before_action :load_form_answer
+    base.before_action :load_feedback, except: :create
+  end
+
+  def create
```
Comment (matthewford) [inline app/concerns/controllers/feedback_mixin.rb]: Assignment Branch Condition size for create is too high. [16.06/15]

File: app/concerns/controllers/feedback_mixin.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        end
+
+        redirect_to polimorphic_url([namespace_name, @form_answer])
+      end
+
+      format.js { render "admin/feedbacks/create" }
+    end
+  end
+
+  def update
```
Comment (matthewford) [inline app/concerns/controllers/feedback_mixin.rb]: Assignment Branch Condition size for update is too high. [15.03/15]

File: app/forms/feedback_form.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  def self.fields_for_award_type(type)
+    keys = {}
+    const_get("AppraisalForm::#{type.upcase}").each do |k, v|
+      keys[k] = { label: v[:label] }
+    end
+
+    filtered_fields(keys)
+  end
+
+  private
```
Comment (matthewford) [inline app/forms/feedback_form.rb]: Useless `private` access modifier.

File: app/models/form_answer.rb [hunk 4]
```diff
@@ -45,6 +45,7 @@ class FormAnswer < ActiveRecord::Base
     has_one :development_eligibility, class_name: 'Eligibility::Development', dependent: :destroy
     has_one :promotion_eligibility, class_name: 'Eligibility::Promotion', dependent: :destroy
     has_one :audit_certificate, dependent: :destroy
+    has_one :feedback, dependent: :destroy
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/policies/feedback_policy.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def update?
+    return @can_update unless @can_update.nil?
+
+    @can_update = if assessor?
+      subject.lead?(form_answer) || (subject.assigned?(form_answer) && !record.submitted?)
+    else
+      admin?
+    end
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: `end` at 13, 4 is not aligned with `if` at 9, 18

File: app/policies/feedback_policy.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  alias :create? :update?
+
+  def submit?
+    return @can_submit unless @can_submit.nil?
+
+    @can_submit = if assessor?
+      !subject.lead?(form_answer) && subject.assigned?(form_answer) && !record.submitted?
+    else
+      admin? && !record.submitted?
+    end
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: `end` at 24, 4 is not aligned with `if` at 20, 18

File: app/policies/feedback_policy.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def approve?
+    return @can_approve unless @can_approve.nil?
+
+    @can_approve = if assessor?
+      subject.lead?(form_answer) && record.submitted? && !record.approved?
+    else
+      admin? && record.submitted? && !record.approved?
+    end
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: `end` at 34, 4 is not aligned with `if` at 30, 19

File: app/policies/feedback_policy.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+    @can_update = if assessor?
+      subject.lead?(form_answer) || (subject.assigned?(form_answer) && !record.submitted?)
+    else
+      admin?
+    end
+  end
+  alias :create? :update?
+
+  def submit?
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Assignment Branch Condition size for submit? is too high. [16.79/15]

File: app/policies/feedback_policy.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    return @can_submit unless @can_submit.nil?
+
+    @can_submit = if assessor?
+      !subject.lead?(form_answer) && subject.assigned?(form_answer) && !record.submitted?
+    else
+      admin? && !record.submitted?
+    end
+  end
+
+  def approve?
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Assignment Branch Condition size for approve? is too high. [17.12/15]

File: app/policies/feedback_policy.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  def create?
+    admin? || (assessor? && subject.assigned?(form_answer))
+  end
+
+  def update?
+    return @can_update unless @can_update.nil?
+
+    @can_update = if assessor?
+      subject.lead?(form_answer) || (subject.assigned?(form_answer) && !record.submitted?)
+    else
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Align `else` with `if`.

File: app/policies/feedback_policy.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+  alias :create? :update?
+
+  def submit?
+    return @can_submit unless @can_submit.nil?
+
+    @can_submit = if assessor?
+      !subject.lead?(form_answer) && subject.assigned?(form_answer) && !record.submitted?
+    else
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Align `else` with `if`.

File: app/policies/feedback_policy.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      admin? && !record.submitted?
+    end
+  end
+
+  def approve?
+    return @can_approve unless @can_approve.nil?
+
+    @can_approve = if assessor?
+      subject.lead?(form_answer) && record.submitted? && !record.approved?
+    else
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Align `else` with `if`.

File: app/policies/feedback_policy.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
+class FeedbackPolicy < ApplicationPolicy
+  def create?
+    admin? || (assessor? && subject.assigned?(form_answer))
+  end
+
+  def update?
+    return @can_update unless @can_update.nil?
+
+    @can_update = if assessor?
+      subject.lead?(form_answer) || (subject.assigned?(form_answer) && !record.submitted?)
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Use 2 (not -12) spaces for indentation.

File: app/policies/feedback_policy.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+      admin?
+    end
+  end
+  alias :create? :update?
+
+  def submit?
+    return @can_submit unless @can_submit.nil?
+
+    @can_submit = if assessor?
+      !subject.lead?(form_answer) && subject.assigned?(form_answer) && !record.submitted?
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Use 2 (not -12) spaces for indentation.

File: app/policies/feedback_policy.rb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    else
+      admin? && !record.submitted?
+    end
+  end
+
+  def approve?
+    return @can_approve unless @can_approve.nil?
+
+    @can_approve = if assessor?
+      subject.lead?(form_answer) && record.submitted? && !record.approved?
```
Comment (matthewford) [inline app/policies/feedback_policy.rb]: Use 2 (not -13) spaces for indentation.

File: app/concerns/controllers/feedback_mixin.rb [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+    respond_to do |format|
+      format.html do
+        if @feedback.valid?
+          flash.notice = "Feedback was successfully updated"
+        else
+          flash.alert = "Feedback was not updated"
+        end
+
+        redirect_to polimorphic_url([namespace_name, @form_answer])
```
Comment (jaleszek) [inline app/concerns/controllers/feedback_mixin.rb]: Do we need the `polimorphic_url` here ? Maybe just `redirect_to [namespace_name, @form_answer]` can do the job ?

File: app/concerns/controllers/feedback_mixin.rb [hunk 17]
```diff
@@ -0,0 +1,97 @@
+module Controllers::FeedbackMixin
```
Comment (jaleszek) [inline app/concerns/controllers/feedback_mixin.rb]: What do you think about storing these mixins in app/controllers/concerns ? I guess we don't need to create the /controllers namespace then and focus on wrapping domain into namespace name - I'm not sure but probably we need to unify it.

**Comment by jaleszek** [comment]
@dreamfall looks good, just let me know about your thoughts ^^.

File: app/controllers/concerns/feedback_mixin.rb [hunk 18]
```diff
@@ -0,0 +1,97 @@
+module FeedbackMixin
+  extend ActiveSupport::Concern
+
+  def self.included(base)
+    base.before_action :load_form_answer
+    base.before_action :load_feedback, except: :create
+  end
+
+  def create
```
Comment (matthewford) [inline app/controllers/concerns/feedback_mixin.rb]: Assignment Branch Condition size for create is too high. [15.07/15]

**Pull request merged**